### PR TITLE
feature(Service): inject infinite services without worry

### PR DESCRIPTION
**Pull request by Stradivario**
Inject infinite services without worry if they are with the same name, 4 lines of code changes everything :D

In the current Typedi usage you are not allowed to use ListService two times because it will get the container by name.NOW WE CAN!!! WITH HASHES!! every class name is now hash representing string from the content of the file which is unique so we are allowed to use singleton services even with the same name based on this identifier c81c01f7131a4c1d5bbd9bda340dce5d its fucking beautifull guys!!!! :D :D 

What i have tested is that when i execute `${target}` it converts to raw class content.Which leads to creating unique identifier for every service :D
Based on my experience with typedi there is a singleton pattern but if you have two services named ListService for example users can have list with ListService and product can have ListService.
Inside the same singleton skeleton you are not allowed to do it because of the non unique identifier from the constructor.name or target.name.Now we can! As long as they are with different content :D i am really happy that i found out this but i was working on a solution from a long time ago.

Cheers!!! 

Regards,
Kristiyan Tachev

### CODE REVIEW

**Comment by Stradivario** [comment]
Implemented two more features making architecture even more fun.

OnInit and OnBefore hooks which can be used as follows

```typescript
import { Service, OnBefore, OnInit } from 'typedi';

@Service()
+export class QuestionUtils implements OnBefore, OnInit {
+  
+  constructor() {
+      // 2
+  }
+
+  OnInit() {
+      // 3 
        this represents the same this as constructor has
+  }
+
+  OnBefore() {
+      // 1
        this represents constructor before initialization can be useful if you are changing something inside prototype or constructor before real intialization.Can be used when you are creating some framework with custom decorators using typedi
+  }
+}
```

**Comment by Stradivario** [comment]
Hello guys i have created based on the idea and abstraction of your library @rxdi/core which can be found here: https://www.npmjs.com/package/@rxdi/core it is a beta testing at this moment but it is working as expected.I need to fill the documentation more but anyway it is super power full and beautiful like architecture.There are a lot of features like lazy evaluation of factories, plugins, afterPlugins, beforePlugins resolvers and many many more which in couple of weeks will be available inside documentation.

Here is a github repository https://github.com/rxdi/core
Here are modules related with this repository.
https://github.com/rxdi/main

Today will be deployed rxdi/hapi which will help to use hapi server provided inside dependency injection.

Regards,
Kristiyan Tachev

**Comment by Stradivario** [comment]
whats wrong ? it is based on typedi why confusing faces :) its just a hard fork with some implementations ;) you really can do anything with typedi thank you guys!

File: README.md [hunk 1]
```diff
@@ -794,6 +794,26 @@ const postController = Container.get(PostController);
 console.log(postController);
 ```
 
+### Container hooks:
```
Comment (justinmchase) [inline README.md]: This seems like an unrelated change and should probably be in a different PR.

File: src/ContainerInstance.ts [hunk 2]
```diff
@@ -333,7 +333,14 @@ export class ContainerInstance {
             // need to be injected, and user can use provided container to get instances he needs
             params.push(this);
 
+            if (type.prototype.OnBefore) {
+                type.prototype.OnBefore.bind(type)();
```
Comment (justinmchase) [inline src/ContainerInstance.ts]: It does not seem right to bind a method like this to a Type. Invoking a member method and altering the `this` to be something other than an instance of that Type is naughty.

File: src/ContainerInstance.ts [hunk 3]
```diff
@@ -333,7 +333,14 @@ export class ContainerInstance {
...
+            }
             value = new (type.bind.apply(type, params))();
+
+            if (value.OnInit) {
+                value.OnInit.bind(value)();
```
Comment (justinmchase) [inline src/ContainerInstance.ts]: In this case you can just do `value.OnInit()` there is no need for bind.

File: src/decorators/Service.ts [hunk 4]
```diff
@@ -93,10 +96,15 @@ export function Service<T, K extends keyof T>(optionsOrServiceName?: ServiceOpti
     } else {
         return function(target: Function) {
 
+
+            Object.defineProperty(target, "originalName", { value: target.name || target.constructor.name, writable: false });
+            Object.defineProperty(target, "name", { value: createHash("md5").update(`${target}`).digest("hex"), writable: true });
```
Comment (justinmchase) [inline src/decorators/Service.ts]: You really shouldn't define a property on someone elses object like this it can lead to really subtle bugs for people who have intentionally used the name or originalName properties already. You could safely use a `Symbol()` here instead however.

Then later when resolving the key instead of just using the functions name, look it up by the symbol.

**Comment by pleerock** [comment]
I don't feel like such things should be part of this library.

**Comment by Stradivario** [comment]
@pleerock sad to hear that :)

**Comment by Stradivario** [comment]
Happy that we have fork button indeed :D

**Comment by justinmchase** [comment]
@Stradivario It seems like what you're trying to accomplish with this is to create the same type multiple times in the same Container instance. I just wanted to point out that this is already available in three different ways that may require you to do some refactoring to utilize but would probably be safer and more deterministic.

## Multiple Container Instances
You can simply have multiple container instances and when resolving your service, each container instance will resolve a new object for that type.

```ts
let id = 0
@Service()
class A {
  public id = id++
}
const c0 = Container.of(0)
const c1 = Container.of(1)

console.log(c0.get(A).id) // 0
console.log(c1.get(A).id) // 1
```

## Reset a Container Instance
If you are done with an instance and want a new one, you can reset the type and it will clear it from the cache and will resolve a new instance the next time it is resolved.


```ts
let id = 0
@Service()
class A {
  public id = id++
}
const c = Container.of(0)
console.log(c.get(A).id) // 0
c.reset()
console.log(c.get(A).id) // 1
```

## Use `multiple: true` to have multiple instances of the same service type
You can have a "multiple" service type which will allow you to have multiple instances of different types which can be resolved individually or as a set.

```ts
interface IExample {
  name: string
}

export const ExampleToken = new Token<IExample>("examples");

@Service({ id: ExampleToken, multiple: true })
class A implements IExample {
  public name = "A"
}

@Service({ id: ExampleToken, multiple: true })
class B implements IExample {
  public name = "B"
}

const examples = Container.getMany(ExampleToken);
console.log(examples.map(ex => ex.name)) // ["A", "B"]
console.log(Container.get(A).name) // "A"
console.log(Container.get(B).name) // "B"
```
